lock-api/src/main/java/com/palantir/lock/client/CoalescingTransactionService.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.client;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.palantir.atlasdb.autobatch.DisruptorAutobatcher;
+import com.palantir.common.base.Throwables;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.StartIdentifiedAtlasDbTransactionResponse;
+import com.palantir.lock.v2.StartTransactionResponseV4;
+import com.palantir.lock.v2.TimestampAndPartition;
+
+final class CoalescingTransactionService {
```

Comment:
The more I think about, the more I think that this class needn't exist. We should just autobatch by default inside `LockLeaseService`. The contract is the same, we don't need to test that it will work, since we have tests for autobatcher. Akin to caching.

Once you've moved out all the different misplaced pieces, it actually doesn't do much besides tunnel calls into autobatcher. The same way we wouldn't test a cache (unless that cache is doing different things such as in `CachingPartitionAllocator`), I don't think it's necessary to test autobatcher.

Reply:
This is a slightly wild view, so can do this last, but I think overall it does make things cleaner.

Reply:
I have concerns on adding the coalescing logic to lockLeaseService. I believe it would make it pretty hard to follow, but functionalities are complex enough to have its own place. Of course, this layered approach is not the only way of decomposing the functionality, but in any case I believe we need to separate them in some way. Tried something different here: #3844